shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -14,333 +14,773 @@
     }
+
+    return { vertex: { grain: { _id: grainId, identityId: identityId } },
+             permissions: permissions, };
   }
 
-  if (roleAssignment) {
-    result.add(new PermissionSet(roleAssignment.addPermissionSet));
-    result.remove(new PermissionSet(roleAssignment.removePermissionSet));
+  conjoin(other) {
+    // Updates this clause to include the permissions contained in `other`.
+    check(other, Clause);
+
+    for (let grainId in other.identityPermissions) {
+      for (let identityId in other.identityPermissions[grainId]) {
+        let otherPermissions = other.identityPermissions[grainId][identityId];
+        this._ensureEntryExists(grainId, identityId);
+        this.identityPermissions[grainId][identityId].add(otherPermissions);
+      }
+    }
   }
-  return result;
 }
 
-function collectEdges(db, vertex) {
-  // Given a vertex in the sharing graph in the format specified by the `check()` invocation below,
-  // collects the data needed for permissions computations pertaining to that vertex. There are three
-  // self-explanatory special cases for the return value. In order of decreasing precedence, they
-  // are: `{grainDoesNotExist: true}`, `{grainIsPublic: true}`, and `{disallowedAnonymousAccess: true}`.
-  // In all other cases, this function returns an object of the form:
-  //   `{edgesByRecipient: <object>, terminalEdge: Optional(<object>)}`.
-  // The `edgesByRecipient` field is a map that coalesces chains of `parentToken` UiView tokens into
-  // direct user-to-user edges; its keys are identity IDs of recipient users and its values are lists
-  // of "edge" objects of the form
-  //   `{sharer: OneOf(<identityId>, "OwningAccount"), roleAssignments: <list of role assignments>}`.
-  // The role assignments should be applied in sequence to compute the set of permissions that flow to a
-  // recipient from a sharer. The `terminalEdge` field is an edge object representing the link to
-  // `vertex` from the nearest user in the sharing graph. If `vertex` is already a user, then this edge is
-  // trivial and its `roleAssignments` field is an empty list. If `terminalEdge` is not present,
-  // then there is no such link.
-  //
-  // TODO(now): UiView tokens can have membrane requirements; we need to account for them in this
-  // computation.
-  check(vertex,
-        Match.OneOf({token: Match.ObjectIncluding({_id: String, grainId: String})},
-                    {grain: Match.ObjectIncluding(
-                      {_id: String,
-                       identityId: Match.OneOf(String, null, undefined)})}));
-
-  var grainId;
-  if (vertex.token) {
-    grainId = vertex.token.grainId;
-  } else if (vertex.grain) {
-    grainId = vertex.grain._id;
-  }
-  var grain = db.getGrain(grainId);
-  if (!grain) {
-    return {grainDoesNotExist: true};
-  }
+class MembranedPermissionSet {
+  // A PermissionSet that is contingent upon some membrane requirements. The membrane is repesented
```

Comment:
It took me a bit to understand what you mean by "the membrane is represented as a Clause". What you actually mean is "the requirements are represented as a Clause", right?